compiler/rustc_trait_selection/src/solve/eval_ctxt/canonical/dedup_solver/mod.rs

```diff
@@ -0,0 +1,210 @@
+        to_remove: &FxIndexSet<usize>,
+        constraints: &mut ExternalConstraintsData<'tcx>,
+    ) {
+        let mut remove_outlives: FxIndexSet<usize> =
+            to_remove.iter().filter_map(|x| self.indx_to_outlives.get(x)).cloned().collect();
+        let mut remove_members: FxIndexSet<usize> =
+            to_remove.iter().filter_map(|x| self.indx_to_members.get(x)).cloned().collect();
+        remove_outlives.sort();
+        remove_members.sort();
+
+        for indx in remove_outlives.into_iter().rev() {
+            constraints.region_constraints.outlives.swap_remove(indx);
+        }
+        for indx in remove_members.into_iter().rev() {
+            constraints.region_constraints.member_constraints.swap_remove(indx);
+        }
+    }
+    fn compress_variables(
+        &mut self,
+        removed_vars: &FxIndexSet<usize>,
+        constraints: &mut ExternalConstraintsData<'tcx>,
+        variables: &mut CanonicalVarInfos<'tcx>,
+        max_universe: &mut UniverseIndex,
+    ) {
+        let mut vars = variables.as_slice().to_vec();
+        let mut universes_available: FxIndexSet<UniverseIndex> =
+            vars.iter().map(|x| x.universe()).collect();
+        universes_available.sort();
+
+        let mut compressed_vars: FxHashMap<usize, usize> = FxHashMap::default();
+        let mut universes_used: FxIndexSet<UniverseIndex> = FxIndexSet::default();
+
+        let mut num_removed = 0;
+        let mut var_indx = 0;
+        while var_indx < vars.len() {
+            let original_var_indx = var_indx + num_removed;
+            if removed_vars.contains(&original_var_indx) {
+                num_removed += 1;
+                vars.remove(var_indx);
+                continue;
+            }
+            compressed_vars.insert(original_var_indx, var_indx);
+            universes_used.insert(vars[var_indx].universe());
+            var_indx += 1;
+        }
+        universes_used.sort();
+
+        for var in vars.iter_mut() {
+            *var = var.with_updated_universe(
+                *universes_available
+                    .get_index(universes_used.get_index_of(&var.universe()).unwrap())
+                    .unwrap(),
+            );
+        }
+
+        let mut var_rewriter = |var| compressed_vars.get(&var).cloned().unwrap_or(var);
+        for outlives in constraints.region_constraints.outlives.iter_mut() {
+            let mut walker = ConstraintWalker::new(self.tcx, &mut var_rewriter);
+            outlives.0 = walker.walk_outlives(&outlives.0);
+        }
+        for member in constraints.region_constraints.member_constraints.iter_mut() {
+            let mut walker = ConstraintWalker::new(self.tcx, &mut var_rewriter);
+            *member = walker.walk_members(member);
+        }
+
+        *variables = self.tcx.mk_canonical_var_infos(&vars);
+        *max_universe = UniverseIndex::from(universes_used.len().saturating_sub(1));
+    }
+}
+
+fn dedup_exact_eq<T: PartialEq>(input: &mut Vec<T>) {
```

Comment:
maybe that code should use an `FxIndexMap` instead so that the exact eq deduplication happens automatically?